Refactor optionlist onSelect handlers

nmajor25 opened the pull request:
Shortcut Story ID: [sc-64123]

- [x] Relatively simple refactors
- [x] Refactor add workflow app/views optionlist
- [x] Remove code from optionlist component

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Refactor**
  * Unified option selection across menus to a centralized event-driven flow, replacing per-item callbacks.
  * Standardized triggers for add, edit, delete, and archive actions for more consistent behavior.
  * Reduced reliance on utility bindings for cleaner, more maintainable code.

* **Improvements**
  * More reliable action handling from menus.
  * Bulk select toggles only when needed, avoiding unnecessary state changes.

* **Tests**
  * Updated component and integration tests to align with the new selection flow.
  * Added brief waits to stabilize list updates during test runs.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nmajor25 committed:
Refactor optionlist onSelect handlers

paulfalgout commented on src/js/views/patients/patient/sidebar/sidebar_views.js:
```diff
@@ -100,6 +97,14 @@ const SidebarView = View.extend({
     });
 
+    this.listenTo(optionlist, 'select', ({ model }) => {
+      const optionItem = model.get('key');
```
I think for this one rather than key you just put event and make it
```
event: canEditPatient ? 'click:patientEdit' : 'click:patientView',
...
event: 'click:activeStatus',
...
event: 'click:shouldArchive'
```
then make an `onShouldArchive` that either contains or calls `showConfirmArchiveModal`

nmajor25 replied:
That's a good way to do it 👍

paulfalgout reviewed (changes requested):
One small change.. ~~then I think we should remove the onSelect functionality from the optionlist component~~ I see this in the checklist now

nmajor25 committed:
Refactor add workflow button optionlist onselect

nmajor25 commented on src/js/apps/patients/patient/flow/flow_app.js:
```diff
@@ -146,16 +146,22 @@ export default SubRouterApp.extend({
+      this.triggerMethod('add:programAction', programItem);
+    });
+
+    this.showChildView('tools', addButtonView);
```
This works when I test it myself in my local dev environment.

But it's failing in the cypress test. No idea why yet.

nmajor25 replied:
The `'select'` listener in the `add-workflow_views.js` isn't even be called when an option list item is clicked on in the test.

nmajor25 replied:
But if I pause the test, close the option list myself manually, and then click on an option list item. It works.

paulfalgout replied:
need to figure this out before we merge this

nmajor25 replied:
Still haven't figured out why that test isn't working. But I agree, its a sign that something is off with this.

nmajor25 commented on test/integration/patients/patient/patient-flow.js:
```diff
@@ -1085,6 +1085,15 @@ context('patient flow page', function() {
+    cy
+      .get('.patient-flow__actions')
+      .contains('Add')
+      .click();
```
`this.listenTo(optionlist, 'select'...` doesn't fire in this test when the optionlist is rendered the first time.

Closing the optionlist and then re-opening it makes the listener work. I'm not sure why.

shadowhand replied:
At least for standard `<select>` elements, the `onselect` only fires when the option is changed and does not fire onload.

nmajor25 committed:
Use internal viewEvent in optionlist component for handling on select events

And remove other code no longer needed after adding select listeners in view files

paulfalgout commented on src/js/components/optionlist/index.js:
```diff
@@ -48,15 +48,6 @@ export default Picklist.extend({
   onSelect({ model }) {
     if (model.get('isDisabled')) return;
 
-    const onSelect = model.get('onSelect');
```
I'm not sure this is working as expected.  if we're adding a different select trigger then `isDisabled` isn't doing anything unless we handle it in each event handler.  what we might need to do is trigger an internal event that then does:

```js
if (model.get('isDisabled')) return;
this.triggerMethod('select', model);
this.destroy();
```

nmajor25 replied:
What do you mean by triggering an internal event? 

Is the idea to remove this `onSelect()` all together? Because ya we need to do `this.destroy()` to close the picklist. Which I guess we could do in each `this.listenTo(optionlist...`.

paulfalgout replied:
Like
```js
viewEvents: {
    'picklist:item:select': 'onPicklistSelect',
},
onPicklistSelect({ model }) {
  if (model.get('isDisabled')) return;
  this.triggerMethod('select', model);
  this.destroy();
},
```

nmajor25 replied:
The difference being the use of `viewEvents` instead of `viewTriggers` that's currently being used?

paulfalgout replied:
so the main change is that viewTriggers mean that `onSelect` and the `this.listenTo(myOptionList, 'select'` are the same event handler, so they would _both_ _alway_ need `if (model.get('isDisabled')) return;` at the top.  using viewEvents and triggering `select` means that the trigger only happens if the model isn't disabled and that check doesn't need to be added to every `this.listenTo(myOptionList` handler.

nmajor25 replied:
Okay gotcha, I just updated it to that. Seems to work.

Still need to figure out that test issue.

nmajor25 commented on src/js/apps/patients/patient/flow/flow_app.js:
The test issue has something to do with `showAdd()` being called in `toggleBulkSelect`. Because if I remove this line and do `this.showAdd()` up in the `onStart()`, the test passes. (but that breaks other things).

When the patient flow page is loaded, `toggleBulkSelect()` is called twice. Once via `this.toggleBulkSelect()` and a second time via `this.listenTo(this.editableCollection, 'reset', this.toggleBulkSelect);` when the listview triggers a `change:canEdit`.

This doesn't cause issues in the actual app. But for some reason this is goofing up the test. The `listenTo(optionlist, 'select'...` doesn't trigger after `this.showAdd()` is called a second time.

Still not sure how to resolve this. But I think this is the cause of the issue.

nmajor25 replied:
Note: adding a `cy.wait(200)` to the beginning of the test makes the test pass.

paulfalgout replied:
why does change:canEdit need to be triggered on the initial render?

nmajor25 committed:
Add 'cy.wait(60)' to 'add action' flow app test

To accomodate the 60ms debounce on the listview for triggering 'change:canEdit', which effects visibility of the add action and bulk select buttons

nmajor25 commented on src/js/apps/patients/patient/flow/flow_app.js:
```diff
@@ -77,7 +77,6 @@ export default SubRouterApp.extend({
     }));
 
     this.listenTo(this.editableCollection, 'reset', this.toggleBulkSelect);
-    this.toggleBulkSelect();
```
These are both being called `onStart`, which is breaking the test. 

I'm not sure if this is the correct way to solve this. But we should make sure it's only called once on start somehow.

nmajor25 replied:
I don't think we want to do this. It doesn't break functionality. But it makes the elements load in a different order than they currently are.

nmajor25 replied:
I can think of two options:

1. Jerry-rig the test somehow to wait until after the list is rendered and the `this.listenTo(this.editableCollection, 'reset', this.toggleBulkSelect);` finishes.
2. Add logic to the `showAdd()` that bails out if the button is already being shown. Like:

```js
  showAdd() {
    // NEW LINES ADDED
    const currentView = this.getChildView('tools');
    if (currentView instanceof AddButtonView) {
      return;
    }

    const addButtonView = new AddButtonView({
      headingText: i18n.addActionHeadingText,
      lists: [{ collection: new Backbone.Collection(this.addOpts) }],
    });

    this.listenTo(addButtonView, 'add:programAction', programItem => {
      this.triggerMethod('add:programAction', programItem);
    });

    this.showChildView('tools', addButtonView);
  },
```

paulfalgout replied:
I think maybe the solution is `if (!this.editableCollection.length) this.toggleBulkSelect();` @nmajor25

nmajor25 replied:
One thing with the `if (!this.editableCollection.length) this.toggleBulkSelect();` solution is that it will change how the elements load, causing a layout jump:

https://github.com/user-attachments/assets/a977caa8-d976-4ec0-87a1-3805ba8f7c38
